include/pstl_offload/internal/usm_memory_replacement.h

```diff
@@ -0,0 +1,244 @@
+        __res = __allocate_shared_for_device(__device, __size, __alignment);
+    } else {
+        __res = __get_original_aligned_alloc()(__alignment, __size);
+    }
+
+    if (!__res)
+        return nullptr;
+
+    assert(std::uintptr_t(__res) % __alignment == 0);
+    return __res;
+}
+
+static void* __errno_handling_internal_aligned_alloc(std::size_t __size, std::size_t __alignment) {
+    void* __ptr = __internal_aligned_alloc(__size, __alignment);
+    if (!__ptr) {
+        errno = ENOMEM;
+        return nullptr;
+    }
+    return __ptr;
+}
+
+static void* __internal_operator_new(std::size_t __size, std::size_t __alignment = alignof(std::max_align_t)) {
+    void* __res = __internal_aligned_alloc(__size, __alignment);
+
+    while(!__res) {
+        std::new_handler __handler = std::get_new_handler();
+        if (__handler) {
+            __handler();
+        } else {
+            throw std::bad_alloc{};
+        }
+        __res = __internal_aligned_alloc(__size, __alignment);
+    }
+
+    return __res;
+}
+
+static void* __internal_operator_new(const std::nothrow_t&, std::size_t __size, std::size_t __alignment = alignof(std::max_align_t)) noexcept {
+    void* __res = nullptr;
+    try {
+        __res = __internal_operator_new(__size, __alignment);
+    } catch(...) {}
+    return __res;
+}
+
+} // namespace __pstl_offload
+
+#if __linux__
+
+// valloc, pvalloc, __libc_valloc and __libc_pvalloc are not supported
+// due to unsupported alignment on memory page
+
+extern "C" {
+
+inline void* __attribute__((always_inline)) malloc(std::size_t __size) {
+    return ::__pstl_offload::__errno_handling_internal_aligned_alloc(__size, alignof(std::max_align_t));
+}
+
+inline void* __attribute__((always_inline)) calloc(std::size_t __num, std::size_t __size) {
+    // square root of maximal std::size_t value
+    constexpr std::size_t __overflow_multiplier = std::size_t(1) << (sizeof(std::size_t) * CHAR_BIT / 2);
+    std::size_t __full_size = __num * __size;
+
+    // Check overflow on multiplication
+    if (__num >= __overflow_multiplier || __size >= __overflow_multiplier) {
+        if (__num && __full_size / __num != __size) {
+            errno = ENOMEM;
+            return nullptr;
+        }
+    }
+
+    void* __res = ::__pstl_offload::__errno_handling_internal_aligned_alloc(__full_size, alignof(std::max_align_t));
+    return __res ? std::memset(__res, 0, __full_size) : nullptr;
+}
+
+inline void* __attribute__((always_inline)) realloc(void* __ptr, std::size_t __size) {
+    return ::__pstl_offload::__internal_realloc(__ptr, __size);
+}
+
+inline void* __attribute__((always_inline)) memalign(std::size_t __alignment, std::size_t __size) noexcept {
+    return ::__pstl_offload::__errno_handling_internal_aligned_alloc(__size, __alignment);
+}
+
+inline int __attribute__((always_inline)) posix_memalign(void** __memptr, std::size_t __alignment, std::size_t __size) noexcept {
+    if (!__pstl_offload::__is_power_of_two(__alignment))
```

Comment:
I think we can make one return statement in this function without making readability worse

Reply:
Changed